libraries/AP_HAL_Linux/RCInput_Navio2.cpp

```diff
@@ -0,0 +1,76 @@
+#include <AP_HAL/AP_HAL.h>
+
+#if CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_NAVIO2
+#include <cstdio>
+#include <unistd.h>
+#include <fcntl.h>
+#include <cstdlib>
+
+#include <AP_Common/AP_Common.h>
+
+#include "RCInput_Navio2.h"
+
+using namespace Linux;
+
+extern const AP_HAL::HAL& hal;
+
+#define RCIN_SYSFS_PATH "/sys/kernel/rcio/rcin"
+
+void RCInput_Navio2::init()
+{
+    for (size_t i = 0; i < ARRAY_SIZE(channels); i++) {
+        channels[i] = open_channel(i);
+        if (channels[i] < 0) {
+            perror("open");
+        }
+    }
+}
+
+void RCInput_Navio2::_timer_tick(void)
+{
+    if (AP_HAL::micros() - _last_timestamp < 10000) {
+        return;
+    }
+
+    char buffer[10];
+    uint16_t periods[ARRAY_SIZE(channels)] = {0};
+
+    for (size_t i = 0; i < ARRAY_SIZE(channels); i++) {
+        if (::pread(channels[i], buffer, ARRAY_SIZE(buffer), 0) < 0) {
+            perror("pread");
+            continue;
+        }
+
+        periods[i] = atoi(buffer);
+    }
+
+    _update_periods(periods, ARRAY_SIZE(periods));
+
+    _last_timestamp = AP_HAL::micros();
+}
+
+RCInput_Navio2::RCInput_Navio2()
+{
+
+}
+
+RCInput_Navio2::~RCInput_Navio2()
+{
+}
+
+int RCInput_Navio2::open_channel(int channel)
+{
+    char *channel_path;
+    if (asprintf(&channel_path, "%s/ch%d", RCIN_SYSFS_PATH, channel) == -1) {
```

Comment:
I personally think, dynamic allocation should be avoided. 
Whenever this function is called asprintf will allocate a tiny bit more.